docs: document experimental streaming search API (#15715)
grafana/mimir: docs/sources/mimir/references/http-api/_index.md

{{< /admonition >}}

The streaming search API provides three endpoints that return metric names, label names, and label values matching optional fuzzy search terms and series selectors.
Unlike the [get label names](#get-label-names) and [get label values](#get-label-values) endpoints, which buffer the full result set and respond with a single JSON document, the search endpoints stream their results as [newline-delimited JSON (NDJSON)](https://github.com/ndjson/ndjson-spec): the response is a sequence of JSON objects separated by newlines, allowing a client to begin processing results before the request completes.

These endpoints mirror the experimental search API proposed in Prometheus [pull request #18573](https://github.com/prometheus/prometheus/pull/18573).

These endpoints are disabled by default.
To enable them, set the `-querier.experimental-search-api-enabled` CLI flag (or its respective YAML configuration option) to `true`.
When the feature is disabled, the endpoints respond with HTTP `404 Not Found` and a `feature_not_enabled` error body.

Each endpoint requires [authentication](#authentication).

#### Search metric names

> ```

Reviewer: ```suggestion
```bash
```

GET,POST <prometheus-http-prefix>/api/v1/search/metric_names
```

Returns metric names (the values of the `__name__` label) matching the request's [query parameters](#search-query-parameters).

When `include_metadata=true`, each result can also carry the metric's `type`, `help`, and `unit` metadata, when available.

Metadata is served only by ingesters, which hold it in memory keyed by metric name and sourced from the metadata records included in remote write requests. It isn't served by store-gateways, so metric names returned from long-term storage carry no metadata. As a result, metadata is returned for only a subset of the returned metric names.

#### Search label names

> ```

Reviewer: ```suggestion
```bash
```

GET,POST <prometheus-http-prefix>/api/v1/search/label_names
```

Returns label names matching the request's [query parameters](#search-query-parameters).

#### Search label values

> ```

Reviewer: ```suggestion
```bash
```

- **label** - _mandatory for `search/label_values` only; not used by the other endpoints_ - the name of the label whose values are returned.
- **search[]** - _optional_ - a fuzzy search term to match candidate names or values against. Repeat the parameter to supply multiple terms, which are combined with `OR` semantics. A maximum of 32 terms is allowed per request. When no term is supplied, all candidates that match the other parameters are returned.
- **match[]** - _optional_ - a PromQL series selector that restricts the candidates to those present in matching series. Repeat the parameter to supply multiple selectors, which are combined with `OR` semantics.
- **start** - _optional_ - the start of the time range to search, as a Unix timestamp (in seconds, with optional decimal places) or RFC 3339 timestamp. Defaults to one hour before the current time.
- **end** - _optional_ - the end of the time range to search, in the same formats as `start`. Defaults to the current time. The value must not be before `start`.
- **case_sensitive** - _optional_ - whether `search[]` term matching is case-sensitive. Defaults to `true`.

> - **fuzz_alg** - _optional_ - the fuzzy-matching algorithm used to score `search[]` terms. One of `subsequence` (the default) or `jarowinkler`.

Reviewer: ```suggestion
- **fuzz_alg** - _optional_ - the fuzzy-matching algorithm used to score `search[]` terms. Either `subsequence` (the default) or `jarowinkler`.
```

> - **fuzz_threshold** - _optional_ - the minimum match score a candidate must reach to be returned, as an integer between `0` and `100`. Defaults to `0`.

Reviewer: Maybe add a sentence here to explain which value indicates a closer fuzzy match, because it's unclear how to configure a looser or more strict fuzzy match.  For example, does a value of 100 equal an exact match?

Author: https://github.com/grafana/mimir/pull/15715/commits/ab603c3d6a1f2e2a1d540ca7858bc3a3ddaab57a

> - **sort_by** - _optional_ - the result ordering. One of `alpha` (the default), which sorts alphabetically, or `score`, which sorts by descending relevance score. Using `score` requires at least one `search[]` term.

Reviewer: ```suggestion
- **sort_by** - _optional_ - the result ordering. Either `alpha` (the default), which sorts alphabetically, or `score`, which sorts by descending relevance score. Using `score` requires at least one `search[]` term.
```

> - **sort_dir** - _optional_ - the sort direction when `sort_by=alpha`. One of `asc` (the default) or `dsc`. Cannot be combined with `sort_by=score`.

Reviewer: ```suggestion
- **sort_dir** - _optional_ - the sort direction when `sort_by=alpha`. Either `asc` (the default) or `dsc`. Cannot be combined with `sort_by=score`.
```

- **batch_size** - _optional_ - the maximum number of results carried in each streamed NDJSON batch. Defaults to `100`, and must not exceed `10000`. This parameter controls only the response framing, not the total number of results.
- **include_score** - _optional_ - whether to include the relevance `score` of each result in the response. Defaults to `false`.
- **include_metadata** - _optional_ - whether to attach metric metadata (`type`, `help`, and `unit`) to each result. Only meaningful for the `search/metric_names` endpoint; ignored by the others. Defaults to `false`.

#### Search response format

A successful response has HTTP status `200` and the `Content-Type: application/x-ndjson; charset=utf-8` header.
The body is a sequence of newline-delimited JSON objects.

Result objects are streamed first, each wrapping a batch of up to `batch_size` results in a `results` array:

```json
{ "results": [{ "name": "up" }, { "name": "go_goroutines" }] }
```

For the `search/label_values` endpoint, each result uses a `value` key instead of `name`:

```json
{ "results": [{ "value": "prometheus" }, { "value": "node" }] }
```

> When `include_score=true`, each result also carries a `score` field. When `include_metadata=true` on the `search/metric_names` endpoint, each result can also carry `type`, `help`, and `unit` fields.

Reviewer 2: nit: this is a partial copy of the `Search query parameters` text.

Perhaps:

> Result output can be modified by setting: Search query parameters (link)
